### ISSUE TITLE
XrdPosixDir::nextEntry and dp->d_type

### ISSUE THREAD

**Issue by wyang007** [xrootd/xrootd#2347]
In function XrdPosixDir::nextEntry() (file XrdPosix/XrdPosixDir.cc), the dp->d_type always returns DT_DIR. Should it return DT_UNKNOWN?

**Comment by abh3** [comment]
Indeed, that was an unfortunate choice and should have been DT_UNKNOWN and will be changed for 6.0 as this is an API change which may or may not break things. As an explanation, the reason this works this way is historic, Priot to Linux 2.6.4 (June 25, 2023) the Linux Posix interface did not return the type of entry and so the xrootd interface did not as well. When d_tyype was introduced, the type was set to, erroneously, DT_DIR and we continued not asking the server to return stat() information. Now, of course, we need to supply the directory type as we need to ask the server to include it in the directory listing. We have two options here a) make that the default, or b) implement a new Posix directive that enables it. I am strongly leaning to solution (a) mostly because configuring Posix either via config file or via pss directives is not well documented; e.g.  posix.setop DirlistAll or pss.setopt DirlistAll which enables disting all the directories in a cluster is undocumented. Also, I doubt that enabling this for dirlist will have much of a performance impact. Any other thoughts about this?

**Comment by amadio** [comment]
I think it should not be a problem to change to `DT_UNKNOWN`, since [`readdir(3)`](https://linux.die.net/man/3/readdir) man page says this:

> Currently, only some filesystems (among them: Btrfs, ext2, ext3, and ext4) have full support for returning the file type in d_type.  All applications must properly handle a return of `DT_UNKNOWN`.
